File: sdks/php/src/SakClient.php
<?php

declare(strict_types=1);

namespace SwissArmyNoife\Sdk;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;

/** HTTP admin client (sak335-b). */
final class SakClient
{
    private const DEFAULT_HTTP = 'http://127.0.0.1:8787';

    private string $baseUrl;
    private Client $http;

    public function __construct(?string $baseUrl = null, ?Client $http = null)
    {
        $u = ($baseUrl === null || trim($baseUrl) === '') ? self::DEFAULT_HTTP : $baseUrl;
        $this->baseUrl = rtrim($u, '/');
        $this->http = $http ?? new Client(['timeout' => 30.0]);
    }

    public function baseUrl(): string
    {
        return $this->baseUrl;
    }

    /** @return array<string, mixed> */
    public function health(): array
    {
        return $this->getJson('/health');
    }

    /** @return array<string, mixed> */
    public function listModules(): array
    {
        return $this->getJson('/v1/sak/modules');
    }

    /** @return array<string, mixed> */
    public function getModule(string $id): array
    {
        return $this->getJson('/v1/sak/modules/' . rawurlencode($id));
    }

    /** @return array<string, mixed> */
    public function capacity(): array
    {
        return $this->getJson('/v1/sak/capacity');
    }

    /** @return array<string, mixed> */
    public function listWork(): array
    {
        return $this->getJson('/v1/sak/compute/work');
    }

    /** @return array<string, mixed> */
    public function listNodes(): array
    {
        return $this->getJson('/v1/sak/compute/nodes');
    }

    /**
     * @param array<string, mixed> $body
     * @return array<string, mixed>
     */
    public function computeWork(array $body): array
    {
        return $this->postJson('/v1/sak/compute/work', $body);
    }

    /**
     * @param array<string, mixed> $body
     * @return array<string, mixed>
     */
    public function computeNodes(array $body): array
    {
        return $this->postJson('/v1/sak/compute/nodes', $body);
    }

    /**
     * @param array<string, mixed>|null $payload
     * @return array<string, mixed>
     */
    public function enqueueWork(string $kind, ?array $payload = null): array
    {
        return $this->computeWork([
            'action' => 'enqueue',
            'kind' => $kind,
            'payload' => $payload ?? [],
        ]);
    }

    /** @return array<string, mixed> */
    public function claimWork(string $nodeId): array
    {
        return $this->computeWork(['action' => 'claim', 'node_id' => $nodeId]);
    }

    /**
     * @param array<string, mixed>|null $result
     * @return array<string, mixed>
     */
    public function completeWork(string $workId, string $nodeId, ?array $result = null): array
    {
        return $this->computeWork([
            'action' => 'complete',
            'work_id' => $workId,
            'node_id' => $nodeId,
            'result' => $result ?? [],
        ]);
    }

    /** @return array<string, mixed> */
    public function getWork(string $workId): array
    {
        return $this->computeWork(['action' => 'get', 'work_id' => $workId]);
    }

    /** @return array<string, mixed> */
    public function requeueWork(string $workId): array
    {
        return $this->computeWork(['action' => 'requeue', 'work_id' => $workId]);
    }

    /**
     * @param array<string, mixed>|null $filters
     * @return array<string, mixed>
     */
    public function listWorkFiltered(?array $filters = null): array
    {
        return $this->computeWork(array_merge(['action' => 'list'], $filters ?? []));
    }

    /**
     * @param array<string, mixed>|null $filters
     * @return array<string, mixed>
     */
    public function listNodesFiltered(?array $filters = null): array
    {
        return $this->computeNodes(array_merge(['action' => 'list'], $filters ?? []));
    }

    /**
     * @param list<string>|null $caps
     * @return array<string, mixed>
     */
    public function registerNode(
        string $label,
        ?array $caps = null,
        ?string $nodeId = null,
        ?string $sessionId = null,
    ): array {
        $body = ['action' => 'register', 'label' => $label];
        if ($caps !== null) {
            $body['caps'] = $caps;
        }
        if ($nodeId !== null && $nodeId !== '') {
            $body['node_id'] = $nodeId;
        }
        if ($sessionId !== null && $sessionId !== '') {
            $body['session_id'] = $sessionId;
        }

        return $this->computeNodes($body);
    }

    /** @return array<string, mixed> */
    public function heartbeatNode(string $nodeId): array
    {
        return $this->computeNodes(['action' => 'heartbeat', 'node_id' => $nodeId]);
    }

    /**
     * @return array<string, mixed>
     * @throws GuzzleException
     */
    private function getJson(string $path): array
    {
        $res = $this->http->get($this->baseUrl . $path);
        /** @var array<string, mixed> $decoded */
        $decoded = json_decode((string) $res->getBody(), true, 512, JSON_THROW_ON_ERROR);

        return $decoded;
    }

    /**
     * @param array<string, mixed> $payload
     * @return array<string, mixed>
     * @throws GuzzleException
     */
    private function postJson(string $path, array $payload): array
    {
        $res = $this->http->post($this->baseUrl . $path, ['json' => $payload]);
        /** @var array<string, mixed> $decoded */
        $decoded = json_decode((string) $res->getBody(), true, 512, JSON_THROW_ON_ERROR);

        return $decoded;
    }
}

File: sdks/cpp/src/sak_client.cpp
#include "swissarmynoife/sak_client.hpp"

#include <httplib.h>
#include <stdexcept>

namespace swissarmynoife {
namespace {

void parse_base(const std::string& base, std::string& scheme, std::string& host, int& port) {
  auto u = base;
  while (!u.empty() && u.back() == '/') u.pop_back();
  scheme = "http";
  std::string rest = u;
  if (rest.rfind("https://", 0) == 0) {
    scheme = "https";
    rest = rest.substr(8);
  } else if (rest.rfind("http://", 0) == 0) {
    rest = rest.substr(7);
  }
  auto slash = rest.find('/');
  auto authority = slash == std::string::npos ? rest : rest.substr(0, slash);
  auto colon = authority.rfind(':');
  if (colon != std::string::npos) {
    host = authority.substr(0, colon);
    port = std::stoi(authority.substr(colon + 1));
  } else {
    host = authority;
    port = scheme == "https" ? 443 : 80;
  }
}

}  // namespace

SakClient::SakClient(std::string base_url) {
  while (!base_url.empty() && base_url.back() == '/') base_url.pop_back();
  if (base_url.empty()) base_url = "http://127.0.0.1:8787";
  base_url_ = base_url;
  parse_base(base_url_, scheme_, host_, port_);
}

nlohmann::json SakClient::health() { return get_json("/health"); }
nlohmann::json SakClient::list_modules() { return get_json("/v1/sak/modules"); }
nlohmann::json SakClient::get_module(const std::string& id) {
  return get_json("/v1/sak/modules/" + id);
}
nlohmann::json SakClient::capacity() { return get_json("/v1/sak/capacity"); }
nlohmann::json SakClient::list_work() { return get_json("/v1/sak/compute/work"); }
nlohmann::json SakClient::list_nodes() { return get_json("/v1/sak/compute/nodes"); }

nlohmann::json SakClient::compute_work(const nlohmann::json& body) {
  return post_json("/v1/sak/compute/work", body);
}
nlohmann::json SakClient::compute_nodes(const nlohmann::json& body) {
  return post_json("/v1/sak/compute/nodes", body);
}

nlohmann::json SakClient::enqueue_work(const std::string& kind, const nlohmann::json& payload) {
  return compute_work({{"action", "enqueue"}, {"kind", kind}, {"payload", payload}});
}
nlohmann::json SakClient::claim_work(const std::string& node_id) {
  return compute_work({{"action", "claim"}, {"node_id", node_id}});
}
nlohmann::json SakClient::complete_work(const std::string& work_id, const std::string& node_id,
                                        const nlohmann::json& result) {
  return compute_work(
      {{"action", "complete"}, {"work_id", work_id}, {"node_id", node_id}, {"result", result}});
}
nlohmann::json SakClient::get_work(const std::string& work_id) {
  return compute_work({{"action", "get"}, {"work_id", work_id}});
}
nlohmann::json SakClient::requeue_work(const std::string& work_id) {
  return compute_work({{"action", "requeue"}, {"work_id", work_id}});
}
nlohmann::json SakClient::list_work_filtered(const nlohmann::json& filters) {
  auto body = filters;
  body["action"] = "list";
  return compute_work(body);
}
nlohmann::json SakClient::list_nodes_filtered(const nlohmann::json& filters) {
  auto body = filters;
  body["action"] = "list";
  return compute_nodes(body);
}

nlohmann::json SakClient::get_json(const std::string& path) {
  httplib::Client cli(host_, port_);
  cli.set_connection_timeout(5, 0);
  cli.set_read_timeout(30, 0);
  auto res = cli.Get(path);
  if (!res || res->status < 200 || res->status >= 300) {
    throw std::runtime_error(std::to_string(res ? res->status : 0) + ": " + (res ? res->body : "no response"));
  }
  return nlohmann::json::parse(res->body);
}

nlohmann::json SakClient::post_json(const std::string& path, const nlohmann::json& payload) {
  httplib::Client cli(host_, port_);
  cli.set_connection_timeout(5, 0);
  cli.set_read_timeout(30, 0);
  auto res = cli.Post(path, payload.dump(), "application/json");
  if (!res || res->status < 200 || res->status >= 300) {
    throw std::runtime_error(std::to_string(res ? res->status : 0) + ": " + (res ? res->body : "no response"));
  }
  return nlohmann::json::parse(res->body);
}

}  // namespace swissarmynoife
